R/embed-vertex.R

```
@@ -0,0 +1,160 @@
+#' - `"QUESTION_ANSWERING"`
+#' - `"FACT_VERIFICATION"`
+#' - `"CODE_RETRIEVAL_QUERY"`
+#' For more information about task types, see [Choose an embeddings task type](https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/task-types).
+#' @examples
+#' \dontrun{
+#' embed_google_vertex(
+#'  "hello world",
+#'  model="gemini-embedding-001",
+#'  project = "<your-project-id>",
+#'  location = "us-central1"
+#' )
+#' }
+#'@export
+embed_google_vertex <- function(x, model, location, project_id, task_type = "RETRIEVAL_QUERY") {
+  if (missing(x) || is.null(x)) {
+    args <- capture_args()
+    fn <- partial(quote(ragnar::embed_google_vertex), alist(x = ), args)
+    return(fn)
+  }
+
+  if (is.data.frame(x)) {
+    x[["embedding"]] <- Recall(
+      x[["text"]],
+      model = model,
+      location = location,
+      project_id = project_id,
+      task_type = task_type
+    )
+    return(x)
+  }
+
+  check_character(x)
+  if (!length(x)) {
+    # ideally we'd return a 0-row matrix, but currently the correct
+    # embedding_size is not convenient to access in this context
+    return(NULL)
+  }
+
+  check_string(model, allow_empty = FALSE)
+  credentials <- google_credentials()
+  credentials <- credentials()
+
+  base_req <- vertex_url(location, project_id) |>
+    httr2::request() |>
+    httr2::req_headers(!!!credentials, .redact = names(credentials)) |>
+    httr2::req_url_path_append(
+      "models",
+      paste0(model, ":predict")
+    ) |>
+    httr2::req_error(body = function(resp) {
+      json <- resp_body_json(resp, check_type = FALSE)
+      json$error$message
+    })
+
+
+  out <- list()
+  # The gemini model does not support batches
+  chunk_size <- if (grepl("gemini", model)) 1 else 20
+
+  for (indices in chunk_list(seq_along(x), chunk_size)) {
+    instances <- lapply(indices, \(i) list(task_type = task_type, content = x[[i]]))
+
+    resp <- base_req |>
+      httr2::req_body_json(list(
+        instances = instances
```

Should we include `autoTruncate=FALSE` in the body? Otherwise the input string will be silently truncated before embedding.  

https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/get-text-embeddings

> Each individual input text is further limited to 2048 tokens; any excess is silently truncated. You can also disable silent truncation by setting autoTruncate to false.